module/zcommon/zfs_uio.c

```diff
@@ -80,22 +80,30 @@ uiomove_iov(void *p, size_t n, enum uio_rw rw, struct uio *uio)
 				if (copy_to_user(iov->iov_base+skip, p, cnt))
 					return (EFAULT);
 			} else {
+				unsigned long bytes_left = 0;
 				if (uio->uio_fault_disable) {
 					if (!zfs_access_ok(VERIFY_READ,
 					    (iov->iov_base + skip), cnt)) {
 						return (EFAULT);
 					}
 					pagefault_disable();
-					if (__copy_from_user_inatomic(p,
-					    (iov->iov_base + skip), cnt)) {
-						pagefault_enable();
-						return (EFAULT);
-					}
+					bytes_left = __copy_from_user_inatomic(p, (iov->iov_base + skip), cnt);
 					pagefault_enable();
 				} else {
-					if (copy_from_user(p,
-					    (iov->iov_base + skip), cnt))
-						return (EFAULT);
+					bytes_left = copy_from_user(p, (iov->iov_base + skip), cnt);
+				}
+				if (bytes_left > 0) {
+					unsigned long copied_bytes = cnt - bytes_left;
+					skip += copied_bytes;
+					if (skip == iov->iov_len) {
```

Comment:
Since we know we're handling a partial write I don't think this case can happen.  How about converting it to an assert, `ASSERT3U(skip, <, iov->iov_len)`.

Reply:
Agreed, see comment above.
Added ASSERT3U as requested.